Terraintexture primary edit

unelsson opened the pull request:
Roughly alter terrain data with mouse clicking. Early PR to get some feedback.

Based on https://github.com/OpenMW/openmw/pull/1679
https://youtu.be/caN1x3Mdy7Y?t=23s

unelsson commented:
Implementing proper data altering through commands will take some time, as it's a completely new system I have to learn, and my understanding is still limited. All help is appreciated. I'm currently trying to get some deeper understanding of data structures, looking through code and reading the comments here: https://github.com/OpenMW/openmw/pull/1427

Regarding the data infrastructure, there is something slightly confusing going on, although it might not be a problem. Comparing these two:

https://github.com/OpenMW/openmw/blob/fc72483ddfe1872d4c9450950d03e99beee245e0/apps/opencs/model/world/data.cpp#L552-L553

https://github.com/OpenMW/openmw/blob/fc72483ddfe1872d4c9450950d03e99beee245e0/apps/opencs/model/world/commands.cpp#L44-L45

It looks to me as ltex records (IdTable) are used in dynamic_casting type LandTextureIdTable. https://github.com/OpenMW/openmw/blob/fc72483ddfe1872d4c9450950d03e99beee245e0/apps/opencs/model/world/commands.cpp#L85

Similar casting (IdTable <-> LandTextureIdTable) is used in model/tools/mergestages.cpp and view/world/landcreator.cpp. LandTextureIdTable is based on IdTable and there's only one function CSMWorld::LandTextureIdTable::importTextures that is different.

unelsson commented:
I'm trying to simplify things into command I understand, here's first idea of ModifyLandTexturesCommand. It does change the data, and sometimes undo works, but it's severely bugged so that every other undo command is lost. I'm also thinking how and when should copy command be called (it's good to get proper texture indexes to plugin, but useless after that). If I get this bug fixed, I'll commit to github.

Calling the command from textureeditmode.cpp (undo bug occurs with or without copy command)
```
  QVariant variant;
  variant.setValue(mNew);
  CSMWorld::TouchLandCommand* ltexTouch = new CSMWorld::TouchLandCommand(landTable, ltexTable, cellId);
  CSMWorld::ModifyLandTexturesCommand* ltexCopy = new CSMWorld::ModifyLandTexturesCommand(landTable, ltexTable, cellId, variant);

  document.getUndoStack().beginMacro("Modify records");
  document.getUndoStack().push(ltexTouch);
  document.getUndoStack().push(ltexCopy);
  document.getUndoStack().endMacro();
```

Command
```
CSMWorld::ModifyLandTexturesCommand::ModifyLandTexturesCommand(IdTable& landTable, IdTable& ltexTable,
    const std::string& id, QVariant new_, QUndoCommand* parent)
    : QUndoCommand(parent)
    , mLands(landTable)
    , mLtexs(ltexTable)
    , mId(id)
    , mNewLand(new_)
{
    setText(("Modify " + mId).c_str());
    int textureColumn = mLands.findColumnIndex(CSMWorld::Columns::ColumnId_LandTexturesIndex);
    CSMWorld::LandTexturesColumn::DataType mPointer = mLands.data(mLands.getModelIndex(mId, textureColumn)).value<CSMWorld::LandTexturesColumn::DataType>();
    CSMWorld::LandTexturesColumn::DataType mOld(mPointer);
    mOldLand.setValue(mOld);
    mLands.setData(mLands.getModelIndex(mId, textureColumn), mNewLand);
}

const std::string& CSMWorld::ModifyLandTexturesCommand::getOriginId() const
{
    return mId;
}

const std::string& CSMWorld::ModifyLandTexturesCommand::getDestinationId() const
{
    return mId;
}

void CSMWorld::ModifyLandTexturesCommand::redo()
{
    int textureColumn = mLands.findColumnIndex(CSMWorld::Columns::ColumnId_LandTexturesIndex);
    CSMWorld::LandTexturesColumn::DataType mPointer = mLands.data(mLands.getModelIndex(mId, textureColumn)).value<CSMWorld::LandTexturesColumn::DataType>();
    CSMWorld::LandTexturesColumn::DataType mOld(mPointer);
    mOldLand.setValue(mOld);
    mLands.setData(mLands.getModelIndex(mId, textureColumn), mNewLand);
}
```

unelsson commented:
Hmm.. No particular reason other than getting the job done. I figured from previous discussions that there wasn't an existing command for this, and therefore land and land texture record modifying should have their own commands altogether. Glancing now through the other commands, macro of CopyLandTexturesCommand or TouchLandCommand, followed with ModifyCommand might work. I don't properly understand the relationship between idtables, idcollections, columns, mbase, mmodified, records, universalid's etc., at least not yet.

unelsson commented:
Feature and nice UI is still WIP, but I think reviewing this code in general is helpful and helping me fix bad habits will prevent future bad code. General tips on C++ can be helpful - "don't do this because of ..., do like this instead".

To-do: 
* Implement drag-like texture painting
* Check variable naming according to OpenMW naming convention
* Add some comments (Doxygen style)
* Check if definition of icon and button sizes can be taken from some variable instead of using fixed number TextureBrushWindow::configureButtonInitialSettings
* In BrushSizeControls::BrushSizeControls constructor, think of good settings for Slider. 1-100 is probably a bit excessive.
* Remove unneeded includes
* Check why terminal says "Warning: Unable to find land texture index 0 in plugin 0, using default texture instead" even after TouchLandCommand.
* Check if primaryselectpressed is correct place for opening brush window. If so, where will terrainselection be then, will that need some checks

jvoisin commented:
you can use the `- [ ] My task` and `- [x] My completed task` syntax on github for todo list ;)

zinnschlag commented:
> Hmm.. No particular reason other than getting the job done. I figured from previous discussions that there wasn't an existing command for this, and therefore land and land texture record modifying should have their own commands altogether.

The general idea here is to have as little command classes as reasonably possible. We want to keep the interface through would the data is manipulated small.

> I don't properly understand the relationship between idtables, idcollections, columns, mbase, mmodified, records, universalid's etc., at least not yet.

If you have any specific questions, I am happy to answer them. But explaining these concepts as a whole would be a bit too lengthy.

> Should primary edit brushes affect only one cell (as it's now), or should brush also affect nearby cells

All cells.

I'll have a quick look over the code now to see if there are any obvious problems (like bad C++). If I find any I'll comment inline. I'll do a more thorough review later (probably within the next few days).

zinnschlag commented on apps/opencs/view/render/terraintexturemode.cpp:
```diff
@@ -0,0 +1,328 @@
+#include "../../model/world/universalid.hpp"
+
+
+#include <components/esm/loadland.hpp>
```
We normally include components before anything from the apps directory. And stuff from the local directory (the one the current file is in) is typically included last (except for including the matching hpp file in a cpp file of course; that always comes first). Functionally it doesn't really matter, but it makes the code a bit easier to read.

zinnschlag commented on apps/opencs/view/render/terraintexturemode.cpp:
```diff
@@ -0,0 +1,328 @@
+  button->setCheckable(true);
+}
+
+void CSVRender::TextureBrushWindow::getBrushTexture(std::string brushTexture)
```
Why is this function called get? Shouldn't it be set instead?

zinnschlag commented on apps/opencs/view/render/terraintexturemode.cpp:
```diff
@@ -0,0 +1,328 @@
+    if(buttonPoint->isChecked()) mBrushShape = 0;
+    if(buttonSquare->isChecked()) mBrushShape = 1;
+    if(buttonCircle->isChecked()) mBrushShape = 2;
+    if(buttonCustom->isChecked()) mBrushShape = 3;
```
This whole construct looks suboptimal. But I need to take a closer look before I can give any advice on it.

zinnschlag commented on apps/opencs/view/render/terraintexturemode.cpp:
```diff
@@ -0,0 +1,328 @@
+    std::size_t spacelocation = mCellId.find(space);
+    std::string slicedX = mCellId.substr (hashlocation+1, spacelocation-hashlocation);
+    std::string slicedY = mCellId.substr (spacelocation+1);
+    CSMWorld::CellCoordinates mCoordinates(stoi(slicedX), stoi(slicedY));
```
This is a very roundabout way of parsing a cell ID. I would just use an std::istringstream. There are probably several examples in the code base.

zinnschlag replied:
CSMWorld::CellCoordinates CSMWorld::RegionMap::getIndex for example.

zinnschlag commented on apps/opencs/view/render/terraintexturemode.cpp:
```diff
@@ -0,0 +1,328 @@
+    std::string mBrushTextureInt = mBrushTexture.substr (hashlocation+1);
+    int brushInt = stoi(mBrushTexture.substr (hashlocation+1))+1; // All indices are offset by +1
+
+    if (mBrushShape == 0) mNew[yInCell*landTextureSize+xInCell] = brushInt;
```
A better way to handle this would be via Brush classes (with a shared brush base class).

unelsson replied:
A new class that would have functions like getShape, getSize, setSize, setShape etc.? That would probably affect many parts of the code, and I regarding other editmodes, I wonder if this base class inheriting would be something like BrushBase, TextureEditBrush : BrushBase, LandShapeEditBrush : BrushBase etc.? I guess size and shape are only common things needed for all brushes. Another way to handle this would be to have all brush settings mashed into one class, and therefore have lots of unnecessary variables, but that's probably not an optimal solution.

Handling a Brush class like that also makes me a bit cautious (slower to code) as there's more to think of regarding RAII and dynamic memory allocation. I've tried to stay on RAII-side and use automated memory handling as much as possible, but I wonder if this is a case where custom destructors and copy constructors will be needed.

zinnschlag commented on apps/opencs/view/render/terraintexturemode.cpp:
```diff
@@ -0,0 +1,328 @@
+    CSMWorld::CommandMacro macro (document.getUndoStack(), "Edit texture records");
+    QModelIndex index(landTable.getModelIndex (mCellId, landTable.findColumnIndex (CSMWorld::Columns::ColumnId_LandTexturesIndex)));
+
+    CSMWorld::TouchLandCommand* ltexTouch = new CSMWorld::TouchLandCommand(landTable, ltexTable, mCellId);
```
Please don't do this. Push the Command directly to the macro. With your code, if the second constructor throws an exception, you have a resources leak.

unelsson replied:
I did fix this already, but I'd be interested on deeper explanation on what happens with exception, and why it's a memory leak. TouchLandCommand is based on Qt's QUndoCommand and I've understood it wouldn't need a delete anyway. I'd guess it has something to do with new not matched with delete anyway, but how come it isn't a leak if "new CSMWorld::TouchLandCommand(landTable, ltexTable, mCellId)" is directly pushed to macro?

zinnschlag commented on apps/opencs/view/render/terraintexturemode.hpp:
```diff
@@ -0,0 +1,137 @@
+            QLabel *selectedBrush;
+            QGroupBox *horizontalGroupBox;
+            int mButtonSize;
+            int mBrushSize = 0;
```
Its not a good idea to initialise variables here if you also have a constructor, because that separates the initialisation into two different parts of the code (two different files actually) which makes it harder to verify that everything got initialised correctly.

zinnschlag commented on apps/opencs/view/render/terraintexturemode.cpp:
```diff
@@ -245,13 +251,12 @@ void CSVRender::TerrainTextureMode::handleDropEvent (QDropEvent *event) {
+    char ignore;
+    int cellX = 0;
+    int cellY = 0;
+    streamCoord >> ignore >> cellX >> cellY;
```
On a second thought, that isn't ideal either: We have the CSMWorld::CellCoordinates class, which makes the whole task much easier (fromId function to create CellCoordinates and then getX/getY). I always forget that we have this class (it was a late addition). Sorry about the confusion.

unelsson replied:
No problem, I'm grateful for learning new stuff along the way.

unelsson replied:
Heh, well, std::pair was a new acquaintance, but here's another solution. The question is - is it better? Could/should that be done without extra variables?

```
std::pair<CSMWorld::CellCoordinates, bool> cellCoordinates_pair = CSMWorld::CellCoordinates::fromId (mCellId);
    CSMWorld::CellCoordinates cellCoordinates = cellCoordinates_pair.first;

    int cellX = cellCoordinates.getX();
    int cellY = cellCoordinates.getY();
```

unelsson commented:
> Should primary edit brushes affect only one cell (as it's now), or should brush also affect nearby cells

> All cells.

This will require some fairly big changes to the code, CSVRender::TerrainTextureMode::editTerrainTextureGrid will not look the same after this.

Meta code is something like this:
```
    /* Algorithm description
    This algorithm is only needed with BrushShape 1 (circle) and 2 (square). Case 1 might be
    unnecessary, as Case 2 should be able to handle everything.

    Case 1, when (mBrushSize/2 <= landTextureSize) it can reach only adjancent cells

    check if corner points of square brush go over cell boundary
    if
    xInCell - mBrushSize/2 < 0  or yInCell - mBrushSize/2 < 0 or
    xInCell + mBrushSize/2 > 15 or yInCell - mBrushSize/2 > 15

    then run Case 2 algorithm
    else Do simple one cell data edit

    Case 2, when (when mBrushSize/2 > landTextureSize), it can affect cells farther than adjancent cells

    check the cell coordinates in end points of square brush
                     -mBrushSize/2, -mBrushSize/2
                      mBrushSize/2,  mBrushSize/2

    load all cells in the box into memory, go through data in it and change terrain grid affecting that cell*/
```

unelsson commented:
There's currently no check in terraintexturemode.cpp for checking whether the cell exists (it tries to load and modify cells that aren't there). In addition, there are still some bugs in brush calculations.

zinnschlag commented:
Regarding cells: That is an interesting problem. In OpenMW-CS we intentionally let the user decide which cells to include in a scene view (quite different from what the original CS did). The intention here was to limit editing to the visible cells.
To support this principle within the texture editing we actually need a stronger check. Not only limit editing to existing cells, but to cells included in the scene view.

zinnschlag commented:
> Instance editing mode has a different behavior: If drop is outside a cell, cell will be loaded to scene view. If cell doesn't exist, it creates a new cell.

We actually need to lock this function away behind a user setting. Depending on what kind project the user is working on not occidentally changing cells that are not supposed to be changed is very important.

If we have a user setting for this function (switching it on/off) we can as well use the same user setting for land textures.

Regarding B) I think I answered that already on the forum and on the other (now closed) PR. Let's try not to split up discussion across several places, please.

> model/world/refcollection.cpp has CSMWorld::RefCollection::getNewId, that's used in instance editing mode. I wonder if similar is needed for idtable now?

Doesn't sound reasonably to me, because such a function would be limited to land texture tables. Of course there needs to be a function to find an unused ID, but it certainly does not belong into a general purpose class like IdTable.

unelsson commented:
I'm trying to achieve a check for unused texture id's, but I'm running into problem with idTable.isDeleted(const std::string& id) function, as invalid id's throw exceptions, because function uses getRecord(id). I'd love to get true instead of exception. Is this intended behavior in idTable? I can't think of anything else than using "try" here, but I feel there might be a better option.. Any ideas?

When ltexTable is idTable for land textures:
```ltexTable.isDeleted(newId)```
-> exception

```bool CSMWorld::IdTable::isDeleted (const std::string& id) const
{
    return getRecord (id).isDeleted();
}```

zinnschlag commented:
Regarding the exception: It would be relatively easy to avoid if you worked on collection level instead of table level. Just use searchId. If it does not return -1, access the record and call the isDeleted function.

zinnschlag commented:
Beginning the review now. Will comment inline if necessary.

First thing I noticed is the settings menu. You added the two settings into Scene Drops. I understand why you did that, but they don't match the title of this page (also the title should be 3D Scene Drops, to be consistent).

It seems the whole scene related part of the user settings needs some cleanly (probably splitting off camera stuff from 3D Scene Input into a separate category and some other changes). We can leave it as it is for now but this needs to be fixed at some point. Please let me know if you want that task (after we are finished with this PR) or if I should write a separate issue about it so that someone else can pick it up.

zinnschlag commented on apps/opencs/view/render/terraintexturemode.hpp:
```diff
@@ -0,0 +1,158 @@
+
+        public:
+            BrushSizeControls(const QString &title, QWidget *parent);
+            QSlider *brushSizeSlider;
```
This does not match our naming conventions.

zinnschlag commented:
The texture resource (non-land-texture) drop is not implemented yet it seems. Did you get around to investigate the index-collision issue for the touch command? This is not only an issue for new texture resources. If index collisions aren't handled then the simple act of editing the terrain with existing land textures could corrupt the content file (i.e. two independent dependencies both with their own land texture records).

zinnschlag commented:
And I see that the button mixup is still in place. That is a major problem. I can't even test this (without screwing up my user settings). Never hardcode mouse buttons! For scene editing we have primary edit, secondary edit, primary select and secondary select and the drags and that's it. Further buttons are not available.

This needs to be addressed before I can complete the review. Pausing the review now.

unelsson commented:
Thanks for pointing this problem out. I'll try to find a solution for this. The most difficult part is in [scenetoolmode.cpp](https://github.com/unelsson/openmw/blob/94c6eaebb4b3ff4faff41280209aa858f4b3364b/apps/opencs/view/widget/scenetoolmode.cpp#L140). So far the only way I know to open brush window from modebutton is to change the scenetoolmode.cpp to pass mouse clicks, and I couldn't find a way to point to primaryselect button. By default, left clicks have to be able to change the editmode, and primaryselect has to connect to terraineditmode to open brush window. Some clicks were previously hardcoded in scenetool.cpp and pushbutton.cpp, so I wonder if they are also problematic.

unelsson commented:
Well, the toolbar button is the only place where buttons are hardcoded in the current code. The functions in scene view follow the primary/secondary edit logic, though I haven't tested that with anything other than default config. The implementation for brush settings window is currently hardcoded MidButton from toolbar button, to open up a brush button menu, should that still be changed to shortcut, even if buttons otherwise are hardcoded?

zinnschlag commented:
I'll try to explain once again:
- In the scene view widget you have only the mapped mouse input I listed before. Here you can never use hard-coded buttons and you can not anything beyond that list
- In control elements (like the toolbar buttons) the normal rules for mouse buttons applies: left click performs the primary function, right click opens a menu or some other secondary list or element

For this feature, when selecting the texture editing mode, a new button for texture/brush management needs to be added to the toolbar. Left clicking on that button will open the brush window (transient, in a suitable position). Right clicking will open a list of recent brush configurations. Dragging a texture onto this button will select it for use in the scene.

Later in addition to the brush button, a selection button needs to be added. This is very similar to how object editing is performed.

unelsson commented:
Ok, I've given this some thought. Implementation of brush settings window isn't currently in the place you asked. Doing that will require creation of new class, something like scenetoolrun, but with different function. If that's added, and it's used with hardcoded Qt::LeftButton, it can't have a submenu (like scenetoolmode), and therefore creating a new selection button will likely require another new widget type.

There still are zero button hardcodings regarding scene editing in scene view widget. [Worldspacewidget uses primaryedit](https://github.com/unelsson/openmw/blob/94c6eaebb4b3ff4faff41280209aa858f4b3364b/apps/opencs/view/render/worldspacewidget.cpp#L658) and [only hardcoding is at scenetoolmode.cpp](https://github.com/unelsson/openmw/blob/94c6eaebb4b3ff4faff41280209aa858f4b3364b/apps/opencs/view/widget/scenetoolmode.cpp#L140). I understand what you mean with hardcoded and not-hardcoded, but this problem doesn't exist with the current code.

The texture resource (non-land-texture) drop is implemented [here](https://github.com/unelsson/openmw/blob/94c6eaebb4b3ff4faff41280209aa858f4b3364b/apps/opencs/view/render/terraintexturemode.cpp#L282-L291).

I have no idea what you are talking about regarding 3D Scene Drops or Scene Drops. I haven't touched anything regarding dropping something to the scene. Dragging the brush in the scene is the only thing that's implemented (primary edit).

Although some feedback is on point, I feel like there's something weird about feedback addressing something that maybe was there, but isn't there anymore.. I'm looking at the github difference, and that's the correct version I'm talking about. I wonder if there's some version mismatch somewhere, broken git, something else? Misspelled branch (my github might have old branches that aren't up to date)?

zinnschlag commented:
Hm, it may be that there was something out of sync somewhere. I can't find the issue in this PR's code anymore, but it was definitely not working properly when I was testing locally and that was matching what I was seeing in the code.

> Is creating a new widget class for brush buttons a good idea? 

Yes.

> Should that be a base class for all editmodes that need a brush settings button to toolbar?

Considering that there aren't many edit modes that will need a brush button and that these brush buttons will be very different from each other, this does not seem like a good idea.

zinnschlag commented:
That looks pretty good. Need to do a bit more testing before I am willing to merge. Meanwhile, a few remaining issues:

* Missing tooltips: The brush button needs one (see the toolbar buttons for an example) and the brush shape buttons also should get a simple tooltip (including a statement that the last button isn't implemented yet)
* The layout of the brush window has a bit too much empty space in it. At the very least the top border is too high. But the space between the button row and the size slider could also be reduced a bit.
* Not liking the range limit much. On the slider it is okay, but the numerical widget should allow for something higher (sometimes you do need to change a large amount of land at once). Since having different limits on the slider and the number widget would probably confusing, maybe we can make the upper limit configurable via a user setting?
* The text at the bottom isn't all that useful. Your average user is probably not aware of the texture index he is using. Maybe use the texture name instead?
* The right click window for the button (listing recently used brushes/textures) is still missing. Remember that we already have such a window (on the run button). At the very least the brush list should match it in looks and functionality, but there is room for code re-use.

zinnschlag commented:
Also, it seems we have a merge conflict.

unelsson commented:
Missing texture should be there already with newicons PR that you just merged. Merge conflict is related to that too, simple to fix. I was waiting what happens with it before fixing these. Good points on other things too, I'll see to these fairly soon.

unelsson commented:
Simple fix just ended up broken, so I did a rebase. Tried to test this PR by git cloning OpenMW and then building, but cmake fails with both master and terraintexturemode_primaryedit branch. CMake fails with not finding ffmpeg-config.cmake or FFmpegConfig.cmake.

unelsson commented:
There's an issue regarding the resizing of brush size menu and debug profiles menu. Quite frustrating.

unelsson commented:
- [x] Missing tooltips -> Done, added
- [x] Recently used brushes/textures -> Done
- [x] Brush Window Layout -> Size has been reduced a bit.
- [ ] Setting naming "Scene drops" -> I'd suggest combining scene drop settings and the rest of terrain Editing settings to something like "3d Scene Editing"
- [ ] Brush size limit -> Will implement a user setting for maximum brush size.
- [ ] Selected brush -text -> Id or nickname? I tried having both instead of just id, but it feels cluttered. Having only texture nickname feels lacking, especially when it's possible to have textures with empty nicknames. I somehow like the texture id too, as it's always there and simple, but there are advantages to nickname too, especially with new brush history menu. I think the best idea so far would be to use nicknames, but when that's empty, print id as "fallback behavior"?

unelsson commented:
AppVeyor builds seem to be failing all over, I don't think the problem is in this PR's code.

zinnschlag commented:
Looks good. I'll see that I can get the (hopefully) final review and test done this weekend.

zinnschlag commented:
Crap. The whole terrain texture system seems to be broken (maybe pre-dating your changes). Steps to reproduce: Create a new addon based on Morrowind.esm. When opening the cell view and selecting the terrain texture editing tool, I get:
> An exception has been caught: invalid ID: L0#0

Also, it seems the TerrainTexture table is empty, which is probably the cause of the exception (btw. having no terrain texture present should be handled more gracefully, because that is the default state for a fresh omwgame file IIRC).

Everything else seems to be working fine. I have one minor issue with the right click brush list. It doesn't match the one we are already having in use (doesn't look the same and lacking the remove button). With the fundamental texture issue there were some planned tests I couldn't run, but I haven't been able to find any other problem so far.

Not entirely sure how to proceed from here. It doesn't make a lot of sense to merge a terrain texture editing feature when terrain textures are broken to the point of being useless. Would you be willing to investigate this issue?

I suspect the way we are handling terrain textures from content files other than the current one is incorrect. It seems we are suppressing all records not from the current file. That is probably ill-conceived. We still need to filter somehow, but what we are currently doing is too aggressive. Maybe only filter out land texture records that have not been superseded by other land texture records further down the content file list? This would need further investigation.

unelsson commented:
I did run into this problem earlier, but I thought touching land records solved this at least so that terrain textures could be worked on. I've successfully created a full new world with opencs using my custom github land shape edit, terrain texture edit and object paint. The fundamental issue is still not resolved though, and I'm quite clueless how the problem should be approached. I'll see if I can think of something.

unelsson commented:
I'm thinking that recent change to code that shows the filename of currently selected texture results in exception when terrain texture table is missing. The problem wasn't so clearly visible when ID stored in mBrushTexture was only used to dig up the index number of painted texture. There still were errors in some cases, but no exceptions that would fundamentally break the whole editmode. If some texture was missing, engine would use "default" textures, and one could easily fix the problems by either dragging textures from assets or manually adding the textures that one wanted to use. I assumed this was just the way the data structures was intended to work, and went on developing fixes and more urgent parts of the code.

To resolve errors with missing land textures, I decided to use touch command to import land texture tables from underlying content files. This used to work, as this would automatically result in game fetching the textures used in cells in underlying content files. I can't remember if this would import even the newly used textures... probably not, I think engine would just use default textures, and user would have to manually add textures or drag more from assets.

I still wonder what's the best way of handling exception with invalid texture ID L0#0. Auto-adding default texture is an option, so is defaulting to something like "invalid texture" when L0#0 isn't found (and denying all edits before a valid texture is selected). These are superficial fixes that should fix worst issues.

But how about fixing the underlying deeper data structure issues - where should I look for this filtering or stuff regarding content file layers? Does this affect OpenMW too, or is it just OpenCS? If OpenCS would understand land texture records from underlying content files, should the records even be touched at all? This (again) feels like a task way over my skill and knowledge, so help is cool!

I played some time with brush history menu layout, and although the layout's fundamentals are copied from debug-profiles menu, it's not exactly the same. That's the best UI I could make so far. I tried using remove-button, but I removed it as I couldn't find a layout where it would be clear - the way it works in debug profiles menu was saying to me "delete the first row of land texture history" instead of "close this window". I'm not against improving it, but I'm not going to make it a big priority.

I also found another bug - when quickly dragging mouse out of cells, or starting drag from outside of land mass, worldspacehit will hit nothing, but textures will still be edited in wrong cell. I think the fix is to just skip land editing when hits don't hit anything - how to find that out, I don't know.

zinnschlag commented:
Wow, lots of stuff. I'll try to provide answers as best as I can. Feel free to follow up on any of my comments.

> I still wonder what's the best way of handling exception with invalid texture ID L0#0. Auto-adding default texture is an option, so is defaulting to something like "invalid texture" when L0#0 isn't found (and denying all edits before a valid texture is selected). These are superficial fixes that should fix worst issues.

Having no land textures is a valid state for a content file. In this case the texture editing tool can not be used, but it should also not fall over with an exception.

The question of the default land texture (implicitly applied by Morrowind/OpenMW) is another question. One I don't have a clear answer to. May it would make sense to use this as a default if no texture exists. But I am not familiar enough with the details to give you any advice on this.

> The fixes above will probably require accessing data through idcollection level instead of idtables. This can be tricky, I don't understand these data structure levels properly, understanding just a little bit of idcollections requires research.

The collections are pretty simple data structures. Just a vector of Records plus an indexing map for quick ID lockup. The records contain the base and the modified part and a state enum and that is really all there is to it.  And just in case (because that apparently confused you in the past): Ignore the columns!

> But how about fixing the underlying deeper data structure issues - where should I look for this filtering or stuff regarding content file layers? Does this affect OpenMW too, or is it just OpenCS? If OpenCS would understand land texture records from underlying content files, should the records even be touched at all? This (again) feels like a task way over my skill and knowledge, so help is cool!

AFAIK this is only an OpenMW-CS problem and yes, we definitely need to work with land texture records. We are editing them after all.

> I played some time with brush history menu layout, and although the layout's fundamentals are copied from debug-profiles menu, it's not exactly the same. That's the best UI I could make so far. I tried using remove-button, but I removed it as I couldn't find a layout where it would be clear - the way it works in debug profiles menu was saying to me "delete the first row of land texture history" instead of "close this window". I'm not against improving it, but I'm not going to make it a big priority.

The remove button is supposed to remove the respective entry from the list. Seems pretty straight forward to me. I don't understand where you have problems with this.

> There are tons of question regarding the data structures.

> Where is texture looked for in following cases? Where it should be looked for? What is Morrowind base files are present? What if they are not?

> What's the behavior/what should it be, when worldspacewidget is painted with texture, when...

The terrain in a content file can only use its own the terrain texture records. If a content file x tries to render its terrain for all intents and purposes the terrain texture records from other content files than x do not exist. Therefore from a rendering perspective there can be no collisions and there are no special cases.

From an editor perspective we need to provide all relevant land texture records. But when using one (to apply it via editing) from a content file that is not the current one we need to make a new record in the current content file and use that instead. The tricky part here seems to be what is relevant? We could just brute force it and show all land texture records in the table. But that would result in poor usability. Still, if everything else fails, that might be an option.

unelsson commented:
I'd really prefer a system that doesn't depend on Morrowind, but rather make this compatible with all games that can be developed for this engine.

I dont see use for removing single entries from brush history, but if that is needed, it's very simple to do.

IdCollections were unfamiliar, but luckily indeed not that difficult. I have hard time reading code using templates, but that's just lack of experience.

I was able to fix the exception, but I'll go on and deny editing when using invalid land textures and probably implement copying of missing land texture record from underlying content files... eg. when painting with L0#0, opencs checks if L0#0 is present in current plugin, if not, it's copied from underlying data, if it doesnt exist anywhere, editing is denied and that is shown in brush selection.

unelsson commented:
The latest commit now handles cases where selected texture id's are invalid, therefore exceptions should be fixed now. The first landtexture-change in a mod is still a bit tricky - this requires either dragging textures from assets or touching a cell that one wants to edit. I don't know what's a better way to do this though, as forcing all landtextures into plugin feels a bit overkill, and so does an extra menu/window that opens up all land textures from underlying layers. I also decided not to implement copying painted texture from underlaying layers, as there is no way to select textures other than valid ones that are already there. So far a way to add textures to plugins is just to drag new textures straight from assets, set index to something not used in underlying layers (eg. 150+), and start painting.

To-do bugfixes still:
- [ ] Missing worldspacehits edit wrong cells
- [ ] Tooltip <br> missing when using invalid textures

zinnschlag commented:
> I'd really prefer a system that doesn't depend on Morrowind, but rather make this compatible with all games that can be developed for this engine.

With Morrowind you mean Morrowind.esm? Misunderstanding, I guess? I never said anything like that.

> I dont see use for removing single entries from brush history, but if that is needed, it's very simple to do.

From a theoretical standpoint: Symmetry. If there is a feature to add something, then there should be a feature to remove that something.

From a practical standpoint: In longer editing sessions the list could become quite long. A user may want to prune it down to the brush settings he still wants to use. Pretty important usability improvement, actually.

> eg. when painting with L0#0, opencs checks if L0#0 is present in current plugin, if not, it's copied from underlying data, if it doesnt exist anywhere, editing is denied and that is shown in brush selection.

That wouldn't work. Keep in mind that the base part is not just a single content file. The same rule applies there: Terrain in a content file can only reference terrain texture records from the same content file. Therefore there could be multiple L0#0 records. If you are painting with L0#0 and there is no such record in the modified section, then something is seriously broken.

When you pick a land texture record from a content file that is not the current one, a new record in the current content file should be created (matching the old one) and then this record should be used.

> The first landtexture-change in a mod is still a bit tricky - this requires either dragging textures from assets or touching a cell that one wants to edit. I don't know what's a better way to do this though, as forcing all landtextures into plugin feels a bit overkill, and so does an extra menu/window that opens up all land textures from underlying layers. I also decided not to implement copying painted texture from underlaying layers, as there is no way to select textures other than valid ones that are already there. So far a way to add textures to plugins is just to drag new textures straight from assets, set index to something not used in underlying layers (eg. 150+), and start painting.

Sorry, but that just won't do. We need the ability to use land texture records from other content files than the current one (via the procedure outlined above). I have done a lot of terrain editing in the vanilla editor and from this experience I can tell you that the lack of this ability would be a showstopper.

The problem here is that the land texture table filters out too many records.

I'll try to sum it up:
- If you edit a cell from another content file all land texture records used in this cell need to be copied to the current content file (with re-indexing, if necessary).
- If a land texture record is picked (for use in a brush) that is not in the current content file, this one record needs to be copied to the current content file (again re-indexing, if necessary).

unelsson commented:
Ok, so.. I've implemented the brute force solution - all ltex records are now shown, but landtexture edit doesn't allow using textures from other plugins. It automatically imports all landtextures used in edited cells to current plugin, and clones a record from underlying layer if user chooses one that's not in current plugin. In case there's a id coillision, a new id is iterated by picking the first free ID in plugin.

Landtexture ID's are L[CONTENTFILEINDEX]#[LANDTEXTUREINDEX]. It seems that plugin index 0 is always the current plugin. The logic in this PR is that if dropped texture is flagged base (is from another plugin), it's then cloned to current plugin and if necessary, reindexed, and newly created texture is then choosed as painting texture. If dropped texture is flagged added, it's considered to be from current plugin and just used like that.

Also, I fixed another bug that occurred when creating a plugin, using brushes that would overlap to neighboring cells. Ltexes would be imported from only the main cell being edited, but not from neighboring cells, resulting in missing textures. This is now fixed, by adding an extra touch command to PushEditToCommand function.

My tests can't break it, but as it's getting more and more complicated, it's hard to keep track of the whole feature.

zinnschlag commented:
Looks good. Will review and test again within the next few days. Hopefully for the last time. Fingers crossed. Who would have thought that texture editing is so complicated? Initially I considered it the easiest part of terrain editing.
